util/kvcache/simple_lru.go

```diff
@@ -73,6 +78,13 @@ func (l *SimpleLRUCache) Put(key Key, value Value) {
 		return
 	}
 
+	var rtm runtime.MemStats
+	runtime.ReadMemStats(&rtm)
+
+	if rtm.Alloc > uint64(float64(l.quota)*(1.0-l.guard)) {
```

Comment:
- `MemQuotaQuery` seems to be a session level variable controlling memory usage for each SQL query, while `runtime.ReadMemStats` get memory usage at `tidb-server` process level, these 2 are not comparable?
- if we simply return when guard ratio is hit, it is inconsistent with virtue of LRU. Shall we remove cache entries from back until we can safely put this entry in?

Reply:
@eurekaka 
(1) I agree that ``MemQuotaQuery`` is not the same level. I should define another configuration to represent the system-wide memory quota or detect the system memory size.
(2) Sure, we can remove the least recently used items until the memory guard condition is met.

Reply:
The updated PR addresses the two comments.